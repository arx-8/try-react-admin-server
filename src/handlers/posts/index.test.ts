import request from "supertest"
import { app } from "../.."
import { excludeUntestableHeaders } from "../../__tests__/testUtils"

describe("/posts", () => {
  const agent = request(app)

  it("Get list", async () => {
    // ## Arrange ##
    // ## Act ##
    const requested = agent.get("/posts")

    // ## Assert ##
    await requested.then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchInlineSnapshot(`
        Object {
          "access-control-allow-headers": "authorization, x-total-count, content-type",
          "access-control-allow-methods": "DELETE, GET, OPTIONS, POST, PUT",
          "access-control-allow-origin": "*",
          "access-control-expose-headers": "authorization, x-total-count, content-type",
          "connection": "close",
          "content-length": "2425",
          "content-type": "application/json; charset=utf-8",
          "etag": "W/\\"979-lrX42kIYjA4vz22AWe2Rm5UiKus\\"",
          "x-powered-by": "Express",
          "x-total-count": "35",
        }
      `)
      expect(resp.body).toMatchSnapshot()
    })
  })

  it("Get list with filters", async () => {
    // ## Arrange ##
    // ## Act ##
    const requested = agent.get(
      "/posts?_start=1&_end=3&_order=DESC&_sort=title"
    )

    // ## Assert ##
    await requested.then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()
    })
  })

  it("Get list with filters by id", async () => {
    // ## Arrange ##
    // ## Act ##
    const requested = agent.get("/posts?id=9")

    // ## Assert ##
    await requested.then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()
    })
  })

  it("Get list with filters by ids", async () => {
    // ## Arrange ##
    // ## Act ##
    const requested = agent.get("/posts?id=1&id=9")

    // ## Assert ##
    await requested.then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()
    })
  })

  it("Get details", async () => {
    // ## Arrange ##
    // ## Act ##
    const requested = agent.get("/posts/2")

    // ## Assert ##
    await requested.then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()
    })
  })

  it("Create", async () => {
    // ## Arrange ##
    // ## Act ##
    const requested = agent.post("/posts").send({
      body: "A new body",
      title: "A new title",
      userId: 2,
    })

    // ## Assert ##
    let id
    await requested.then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()

      id = resp.body.id
    })

    // get again
    await agent.get(`/posts/${id}`).then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()
    })
  })

  it("Update", async () => {
    // ## Arrange ##
    // ## Act ##
    const requested = agent.put("/posts/3").send({
      body: "A edited body",
      id: 3,
      title: "A edited title",
      userId: 3,
    })

    // ## Assert ##
    await requested.then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()
    })

    // get again
    await agent.get("/posts/3").then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()
    })
  })

  it("Delete", async () => {
    // ## Arrange ##
    // ## Act ##
    const requested = agent.delete("/posts/4")

    // ## Assert ##
    await requested.then((resp) => {
      expect(resp.statusCode).toStrictEqual(200)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()
    })
  })

  it("Delete (Not found)", async () => {
    // ## Arrange ##
    // ## Act ##
    const requested = agent.delete("/posts/99999999999999999")

    // ## Assert ##
    await requested.then((resp) => {
      expect(resp.statusCode).toStrictEqual(404)
      expect(excludeUntestableHeaders(resp.headers)).toMatchSnapshot()
      expect(resp.body).toMatchSnapshot()
    })
  })
})
