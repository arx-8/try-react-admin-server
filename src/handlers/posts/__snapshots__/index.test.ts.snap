// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/posts Create 1`] = `
Object {
  "access-control-allow-headers": "authorization, x-total-count, content-type",
  "access-control-allow-methods": "DELETE, GET, OPTIONS, POST, PUT",
  "access-control-allow-origin": "*",
  "access-control-expose-headers": "authorization, x-total-count, content-type",
  "connection": "close",
  "content-length": "9",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"9-rD5kVc2LU759KDT3/yQVEh3R/cQ\\"",
  "x-powered-by": "Express",
}
`;

exports[`/posts Create 2`] = `
Object {
  "id": 36,
}
`;

exports[`/posts Create 3`] = `
Object {
  "access-control-allow-headers": "authorization, x-total-count, content-type",
  "access-control-allow-methods": "DELETE, GET, OPTIONS, POST, PUT",
  "access-control-allow-origin": "*",
  "access-control-expose-headers": "authorization, x-total-count, content-type",
  "connection": "close",
  "content-length": "62",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"3e-bGbPEN+JDQwRoRLm1cSBgSuDCMg\\"",
  "x-powered-by": "Express",
}
`;

exports[`/posts Create 4`] = `
Object {
  "body": "A new body",
  "id": 36,
  "title": "A new title",
  "userId": 2,
}
`;

exports[`/posts Delete (Not found) 1`] = `
Object {
  "access-control-allow-headers": "authorization, x-total-count, content-type",
  "access-control-allow-methods": "DELETE, GET, OPTIONS, POST, PUT",
  "access-control-allow-origin": "*",
  "access-control-expose-headers": "authorization, x-total-count, content-type",
  "connection": "close",
  "content-length": "28",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"1c-i0EptGmraAHmxP2zNRGy7HQtj2g\\"",
  "x-powered-by": "Express",
}
`;

exports[`/posts Delete (Not found) 2`] = `
Object {
  "message": "Data not found",
}
`;

exports[`/posts Delete 1`] = `
Object {
  "access-control-allow-headers": "authorization, x-total-count, content-type",
  "access-control-allow-methods": "DELETE, GET, OPTIONS, POST, PUT",
  "access-control-allow-origin": "*",
  "access-control-expose-headers": "authorization, x-total-count, content-type",
  "connection": "close",
  "content-length": "8",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"8-GmFGyLK2736etq0w88WOBfa7S5o\\"",
  "x-powered-by": "Express",
}
`;

exports[`/posts Delete 2`] = `
Object {
  "id": 4,
}
`;

exports[`/posts Get details 1`] = `
Object {
  "access-control-allow-headers": "authorization, x-total-count, content-type",
  "access-control-allow-methods": "DELETE, GET, OPTIONS, POST, PUT",
  "access-control-allow-origin": "*",
  "access-control-expose-headers": "authorization, x-total-count, content-type",
  "connection": "close",
  "content-length": "261",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"105-YGOc7AML8El062v0h5XH2RJb2Ak\\"",
  "x-powered-by": "Express",
}
`;

exports[`/posts Get details 2`] = `
Object {
  "body": "est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla",
  "id": 2,
  "title": "qui est esse",
  "userId": 1,
}
`;

exports[`/posts Get list 2`] = `
Array [
  Object {
    "body": "quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto",
    "id": 1,
    "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
    "userId": 1,
  },
  Object {
    "body": "est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla",
    "id": 2,
    "title": "qui est esse",
    "userId": 1,
  },
  Object {
    "body": "et iusto sed quo iure
voluptatem occaecati omnis eligendi aut ad
voluptatem doloribus vel accusantium quis pariatur
molestiae porro eius odio et labore et velit aut",
    "id": 3,
    "title": "ea molestias quasi exercitationem repellat qui ipsa sit aut",
    "userId": 1,
  },
  Object {
    "body": "ullam et saepe reiciendis voluptatem adipisci
sit amet autem assumenda provident rerum culpa
quis hic commodi nesciunt rem tenetur doloremque ipsam iure
quis sunt voluptatem rerum illo velit",
    "id": 4,
    "title": "eum et est occaecati",
    "userId": 1,
  },
  Object {
    "body": "repudiandae veniam quaerat sunt sed
alias aut fugiat sit autem sed est
voluptatem omnis possimus esse voluptatibus quis
est aut tenetur dolor neque",
    "id": 5,
    "title": "nesciunt quas odio",
    "userId": 1,
  },
  Object {
    "body": "ut aspernatur corporis harum nihil quis provident sequi
mollitia nobis aliquid molestiae
perspiciatis et ea nemo ab reprehenderit accusantium quas
voluptate dolores velit et doloremque molestiae",
    "id": 6,
    "title": "dolorem eum magni eos aperiam quia",
    "userId": 1,
  },
  Object {
    "body": "dolore placeat quibusdam ea quo vitae
magni quis enim qui quis quo nemo aut saepe
quidem repellat excepturi ut quia
sunt ut sequi eos ea sed quas",
    "id": 7,
    "title": "magnam facilis autem",
    "userId": 1,
  },
  Object {
    "body": "dignissimos aperiam dolorem qui eum
facilis quibusdam animi sint suscipit qui sint possimus cum
quaerat magni maiores excepturi
ipsam ut commodi dolor voluptatum modi aut vitae",
    "id": 8,
    "title": "dolorem dolore est ipsam",
    "userId": 1,
  },
  Object {
    "body": "consectetur animi nesciunt iure dolore
enim quia ad
veniam autem ut quam aut nobis
et est aut quod aut provident voluptas autem voluptas",
    "id": 9,
    "title": "nesciunt iure omnis dolorem tempora et accusantium",
    "userId": 1,
  },
  Object {
    "body": "quo et expedita modi cum officia vel magni
doloribus qui repudiandae
vero nisi sit
quos veniam quod sed accusamus veritatis error",
    "id": 10,
    "title": "optio molestias id quia eum",
    "userId": 1,
  },
]
`;

exports[`/posts Get list with filters 1`] = `
Object {
  "access-control-allow-headers": "authorization, x-total-count, content-type",
  "access-control-allow-methods": "DELETE, GET, OPTIONS, POST, PUT",
  "access-control-allow-origin": "*",
  "access-control-expose-headers": "authorization, x-total-count, content-type",
  "connection": "close",
  "content-length": "459",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"1cb-s+gNSj+z1ehK2xHiVB633IQyJss\\"",
  "x-powered-by": "Express",
  "x-total-count": "35",
}
`;

exports[`/posts Get list with filters 2`] = `
Array [
  Object {
    "body": "eveniet quo quis
laborum totam consequatur non dolor
ut et est repudiandae
est voluptatem vel debitis et magnam",
    "id": 18,
    "title": "voluptate et itaque vero tempora molestiae",
    "userId": 2,
  },
  Object {
    "body": "debitis eius sed quibusdam non quis consectetur vitae
impedit ut qui consequatur sed aut in
quidem sit nostrum et maiores adipisci atque
quaerat voluptatem adipisci repudiandae",
    "id": 31,
    "title": "ullam ut quidem id aut vel consequuntur",
    "userId": 4,
  },
]
`;

exports[`/posts Update 1`] = `
Object {
  "access-control-allow-headers": "authorization, x-total-count, content-type",
  "access-control-allow-methods": "DELETE, GET, OPTIONS, POST, PUT",
  "access-control-allow-origin": "*",
  "access-control-expose-headers": "authorization, x-total-count, content-type",
  "connection": "close",
  "content-length": "8",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"8-0rVkPUSjKXvAXSnOiVTQsEptELY\\"",
  "x-powered-by": "Express",
}
`;

exports[`/posts Update 2`] = `
Object {
  "id": 3,
}
`;

exports[`/posts Update 3`] = `
Object {
  "access-control-allow-headers": "authorization, x-total-count, content-type",
  "access-control-allow-methods": "DELETE, GET, OPTIONS, POST, PUT",
  "access-control-allow-origin": "*",
  "access-control-expose-headers": "authorization, x-total-count, content-type",
  "connection": "close",
  "content-length": "67",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"43-U1p8oEpX5SR+hfHzu+dBYvjQMnE\\"",
  "x-powered-by": "Express",
}
`;

exports[`/posts Update 4`] = `
Object {
  "body": "A edited body",
  "id": 3,
  "title": "A edited title",
  "userId": 3,
}
`;
